src/com/google/enterprise/adaptor/database/DatabaseAdaptor.java

```diff
@@ -723,6 +731,12 @@ static Acl buildAcl(ResultSet rs, String delim, String namespace)
         hasColumn(metadata, GsaSpecialColumns.GSA_PERMIT_GROUPS);
     boolean hasDenyGroups =
         hasColumn(metadata, GsaSpecialColumns.GSA_DENY_GROUPS);
+
+    if (!hasPermitUsers && !hasDenyUsers
+        && !hasPermitGroups && !hasDenyGroups) {
+      return Acl.EMPTY;
```

Comment:
Whether implicit or explicit, the empty ACL is wrong in the case of getting ACLs from db.singleDocContentSql. I think we could just add another argument for the default ACL to use here, either null or Acl.EMPTY. These values are paired with the ResultSetNext enum, so we could conceivably collapse them into a single argument, but that feels a little obfuscated.

Since this is a bug, albeit newly introduced, you should add a test for it.

Reply:
Done.